programs/drift/src/controller/lp.rs

```diff
@@ -26,6 +26,110 @@ use anchor_lang::prelude::Account;
+    let rebase_divisor: i128 = 10_i128.pow(expo_diff.abs().cast()?);
+
+    if expo_diff > 0 {
+        perp_market.amm.base_asset_amount_per_lp = perp_market
+            .amm
+            .base_asset_amount_per_lp
+            .safe_mul(rebase_divisor)?;
+
+        perp_market.amm.quote_asset_amount_per_lp = perp_market
+            .amm
+            .quote_asset_amount_per_lp
+            .safe_mul(rebase_divisor)?;
+
+        perp_market.amm.total_fee_earned_per_lp = perp_market
+            .amm
+            .total_fee_earned_per_lp
+            .safe_mul(rebase_divisor.cast()?)?;
+    } else {
+        perp_market.amm.base_asset_amount_per_lp = perp_market
+            .amm
+            .base_asset_amount_per_lp
+            .safe_div(rebase_divisor)?;
+
+        perp_market.amm.quote_asset_amount_per_lp = perp_market
+            .amm
+            .quote_asset_amount_per_lp
+            .safe_div(rebase_divisor)?;
+
+        perp_market.amm.total_fee_earned_per_lp = perp_market
+            .amm
+            .total_fee_earned_per_lp
+            .safe_div(rebase_divisor.cast()?)?;
+    }
+
+    msg!(
+        "rebasing perp market_index={} per_lp_base expo_diff={}",
+        perp_market.market_index,
+        expo_diff,
+    );
+
+    crate::validation::perp_market::validate_perp_market(perp_market)?;
+
+    Ok(())
+}
+
+pub fn apply_lp_rebase_to_perp_position(
+    perp_market: &PerpMarket,
+    perp_position: &mut PerpPosition,
+) -> DriftResult<()> {
+    let expo_diff = perp_market
+        .amm
+        .per_lp_base
+        .safe_sub(perp_position.per_lp_base)?;
+
+    if expo_diff > 0 {
+        let rebase_divisor: i64 = 10_i64.pow(expo_diff.cast()?);
+
+        perp_position.last_base_asset_amount_per_lp = perp_position
+            .last_base_asset_amount_per_lp
+            .safe_mul(rebase_divisor)?;
+        perp_position.last_quote_asset_amount_per_lp = perp_position
+            .last_quote_asset_amount_per_lp
+            .safe_mul(rebase_divisor)?;
+
+        msg!(
+            "rebasing perp position for market_index={} per_lp_base by expo_diff={}",
+            perp_market.market_index,
+            expo_diff,
+        );
+        perp_position.per_lp_base = perp_position.per_lp_base.safe_add(expo_diff)?;
+    } else if expo_diff < 0 {
+        let rebase_divisor: i64 = 10_i64.pow(expo_diff.abs().cast()?);
+
+        perp_position.last_base_asset_amount_per_lp = perp_position
+            .last_base_asset_amount_per_lp
+            .safe_div(rebase_divisor)?;
+        perp_position.last_quote_asset_amount_per_lp = perp_position
+            .last_quote_asset_amount_per_lp
+            .safe_div(rebase_divisor)?;
+
+        msg!(
+            "rebasing perp position for market_index={} per_lp_base by expo_diff={}",
+            perp_market.market_index,
+            expo_diff,
+        );
+        perp_position.per_lp_base = perp_position.per_lp_base.safe_add(expo_diff)?;
+    }
+
+    Ok(())
+}
+
 pub fn mint_lp_shares(
```

Comment:
dont you need to set the expo in this function when it's not 0 on the perp market?

Reply:
yes, added a test on this. fortunately `validate_perp_position_with_perp_market` caught the issue in mint_lp_shares if that scenario so it would have just blocked